src/main/java/org/springframework/integration/aws/lock/DynamoDbLockRegistry.java

```diff
@@ -0,0 +1,552 @@
+		this.heartbeatPeriod = heartbeatPeriod;
+	}
+
+	public void setRefreshPeriod(long refreshPeriod) {
+		this.refreshPeriod = refreshPeriod;
+	}
+
+	/**
+	 * Set the {@link Executor}, where is not provided then a default of
+	 * cached thread pool Executor will be used.
+	 * @param executor the executor service
+	 */
+	public void setExecutor(Executor executor) {
+		this.executor = executor;
+		this.executorExplicitlySet = true;
+	}
+
+	@Override
+	public void afterPropertiesSet() {
+		if (!this.dynamoDBLockClientExplicitlySet) {
+			AmazonDynamoDBLockClientOptions dynamoDBLockClientOptions =
+					AmazonDynamoDBLockClientOptions
+							.builder(this.dynamoDB, this.tableName)
+							.withPartitionKeyName(this.partitionKey)
+							.withSortKeyName(this.sortKeyName)
+							.withHeartbeatPeriod(this.heartbeatPeriod)
+							.withLeaseDuration(this.leaseDuration)
+							.build();
+
+			this.dynamoDBLockClient = new AmazonDynamoDBLockClient(dynamoDBLockClientOptions);
+		}
+
+		this.leaseDuration =
+				(long) new DirectFieldAccessor(this.dynamoDBLockClient)
+						.getPropertyValue("leaseDurationInMilliseconds");
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return true;
+	}
+
+	@Override
+	public void start() {
+		if (!this.running.getAndSet(true)) {
+			if (!this.dynamoDBLockClientExplicitlySet) {
+				try {
+					this.dynamoDBLockClient.assertLockTableExists();
+					this.createTableLatch.countDown();
+					return;
+				}
+				catch (LockTableDoesNotExistException e) {
+					if (logger.isInfoEnabled()) {
+						logger.info("No table '" + this.tableName + "'. Creating one...");
+					}
+				}
+
+				CreateDynamoDBTableOptions createDynamoDBTableOptions =
+						CreateDynamoDBTableOptions
+								.builder(this.dynamoDB,
+										new ProvisionedThroughput(this.readCapacity, this.writeCapacity),
+										this.tableName)
+								.withPartitionKeyName(this.partitionKey)
+								.withSortKeyName(this.sortKeyName)
+								.build();
+
+				AmazonDynamoDBLockClient.createLockTableInDynamoDB(createDynamoDBTableOptions);
+			}
+		}
+
+		int i = 0;
+
+		boolean tableExists = false;
+
+		// We need up to one minute to wait until table is created on AWS.
+		while (i++ < 60) {
+			tableExists = this.dynamoDBLockClient.lockTableExists();
+			if (!tableExists) {
+				try {
+					// This is allowed minimum for constant AWS requests.
+					Thread.sleep(1000);
+				}
+				catch (InterruptedException e) {
+					ReflectionUtils.rethrowRuntimeException(e);
+				}
+			}
+			else {
+				break;
+			}
+		}
+
+		if (!tableExists) {
+			logger.error("Cannot describe DynamoDb table: " + this.tableName);
+		}
+
+		this.createTableLatch.countDown();
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@Override
+	public void stop() {
+		this.running.set(false);
+	}
+
+	@Override
+	public boolean isRunning() {
+		return false;
+	}
+
+	@Override
+	public int getPhase() {
+		return Integer.MAX_VALUE;
```

Comment:
Why not configurable? Shouldn't it be relatively early?

Reply:
M-m-m. I thought `Integer.MAX_VALUE` is the last phase in the order. No?
Should it be `Integer.MIN_VALUE` then.

On the other hand it really might be better to create table on AWS as early as possible and let our consumers (e.g. aggregator) to do its work just in time.

Reply:
Max value seems wrong in cases where someone wants to get a lock in a Lifecycle.start().

I would have thought earlier than max_value, but certainly not min-value; maybe 0 is the best in this case.